cloud/scope/object_storage_bucket.go

```diff
@@ -117,16 +118,37 @@ func (s *ObjectStorageBucketScope) ApplyAccessKeySecret(ctx context.Context, key
 		return fmt.Errorf("could not set owner ref on access key secret %s: %w", secretName, err)
 	}
 
-	result, err := controllerutil.CreateOrPatch(ctx, s.client, secret, func() error { return nil })
+	if err := controllerutil.SetControllerReference(s.Bucket, secret, s.client.Scheme()); err != nil {
+		return fmt.Errorf("could not set controller ref on access key secret %s: %w", secretName, err)
+	}
+
+	result, err := controllerutil.CreateOrUpdate(ctx, s.client, secret, func() error { return nil })
 	if err != nil {
-		return fmt.Errorf("could not create/patch access key secret %s: %w", secretName, err)
+		return fmt.Errorf("could not create/update access key secret %s: %w", secretName, err)
 	}
 
 	s.Logger.Info(fmt.Sprintf("Secret %s was %s with new access keys", secret.Name, result))
 
 	return nil
 }
 
+func (s *ObjectStorageBucketScope) ShouldInitKeys() bool {
+	return s.Bucket.Status.LastKeyGeneration == nil
+}
+
 func (s *ObjectStorageBucketScope) ShouldRotateKeys() bool {
-	return *s.Bucket.Spec.KeyGeneration != *s.Bucket.Status.LastKeyGeneration
+	return s.Bucket.Status.LastKeyGeneration != nil &&
+		*s.Bucket.Spec.KeyGeneration != *s.Bucket.Status.LastKeyGeneration
+}
+
+func (s *ObjectStorageBucketScope) ShouldRestoreKeySecret(ctx context.Context) (bool, error) {
+	if s.Bucket.Status.KeySecretName != nil {
```

Comment:
small nit, can we inverse the logic here to return early if nil and un nest these conditionals?